### ISSUE TITLE
not working on iphone

### ISSUE THREAD

**Issue by botaoshen** [brentvollebregt/spotify-lyrics-viewer#38]
Hi great app working prefect via pc web-browser. Whereas I am trying to make it work via iphone but seems like not be able to displaying lyric. Understood I can download spotify app to display lyric but the app dosen't showing lyric continuesly which is quite annoying. I am hoping the issue could be addressed

**Comment by brentvollebregt** [comment]
Weirdly I'm unable to even log in using Safari or Chrome on an iPhone I have 🤔. Unfortunately, I'm unable to look at the console to see why this might be happening.

This does seem different to your issue though - it sounds like you are logged in?

**Comment by botaoshen** [comment]
Yes I can sort of passing to the login page but it will not letting me
authorize Spotify via the iOS browsers

Regards,


Brent Vollebregt ***@***.***>于2024年4月19日 周五下午9:12写道：

**Comment by chainespam-stack** [comment]
I’m having the same issue 😭 any luck figuring it out?

**Comment by brentvollebregt** [comment]
@chainespam-stack this issue has not been resolved as I'm still unable to get a debugging setup.

Keen to hear if anyone else has been able to figure out why it's an issue.

**Comment by ghost** [comment]
I have the same issue and i am a dev so since this isnt fixed yet im thinking about forking it and fixing it for you guys

**Comment by botaoshen** [comment]
hi @gl1tter-v1be please keeping us posted

**Comment by ghost** [comment]
> hi [@gl1tter-v1be](https://github.com/gl1tter-v1be) please keeping us posted

Sure i will look into it to as fast i can i think it wil be today in the evening then i can take a look at it

**Comment by botaoshen** [comment]
Hi @gl1tter-v1be have you got this worked out?

**Comment by ghost** [comment]
@botaoshen Not yet but i can look at it today but i need to search for the files that handels the session so i can look at it and fix it

**Comment by brentvollebregt** [comment]
The flow is generally:

1. A user is directed to `/spotify-authorization` which renders the [SpotifyAuthorization](https://github.com/brentvollebregt/spotify-lyrics-viewer/blob/master/client/src/pages/SpotifyAuthorization/index.tsx) page
2. Initially this redirects the user to `/api/spotify/authenticate` on the API side
3. [/api/spotify/authenticate](https://github.com/brentvollebregt/spotify-lyrics-viewer/blob/master/src/routes/spotify.ts#L17) sends the user to Spotify following the authorization code flow
4. Spotify sends the user to [/api/spotify/authentication-callback](https://github.com/brentvollebregt/spotify-lyrics-viewer/blob/master/src/routes/spotify.ts#L49) which redirects the user to where they came from initially with their new access token (using the value stored in `req.session.authentication_origin`) - ideally this is `/spotify-authorization`
5. This time when visiting `/spotify-authorization`, the access token will be in the query string - the value will be stored in the browser

Really this flow itself shouldn't be the issue, but I just haven't been able to obtain a debugging setup to verify that.

I'm happy to review whatever solution you may find and get it into this repo 😃

**Comment by ghost** [comment]
@brentvollebregt It is not due to the authorization flow indeed because I run it on localhost via ubuntu and then I proxy it via a separate extra js file because ngrok requires http and the backend https that is how it works and if I test it via the ngrok domain on my ios phone then I log in successfully so I think the issue lies somewhere else

**Comment by ghost** [comment]
I might have a solution that will hopefully fix it so that it works on iOS via any domain, but I can't promise anything, but the fix I have is iOS-proof, so there's a chance it will fix it.

**Comment by ghost** [comment]
@brentvollebregt I made a pull request https://github.com/brentvollebregt/spotify-lyrics-viewer/pull/44

**Comment by brentvollebregt** [comment]
For what it's worth, I've just discovered it works fine on iPhone when going to https://spotify-lyrics-viewer.onrender.com/ which is where the server is hosted - so it may be a cross origin issue with the cookie.

For the current state of #44, we cannot use "express-session" as the server restarts too much and users would be logged out unexpectedly.

**Comment by brentvollebregt** [comment]
Looking into this more, apparently this is due to iOS-specific cookie restrictions. A cookie being set during a cross-site redirect might be the issue here.

Looks like the only reliable option to fix this is to put the API on a custom domain.

**Comment by ghost** [comment]
> For what it's worth, I've just discovered it works fine on iPhone when going to https://spotify-lyrics-viewer.onrender.com/ which is where the server is hosted - so it may be a cross origin issue with the cookie.
> 
> For the current state of [#44](https://github.com/brentvollebregt/spotify-lyrics-viewer/pull/44), we cannot use "express-session" as the server restarts too much and users would be logged out unexpectedly.

Oh yeah, I hadn't thought about that, Express still has a habit of restarting often, yes I ran it in dev mode so that was my mistake

**Comment by ghost** [comment]
> Looking into this more, apparently this is due to iOS-specific cookie restrictions. A cookie being set during a cross-site redirect might be the issue here.
> 
> Looks like the only reliable option to fix this is to put the API on a custom domain.

That could be right, I ran the API and the frontend via a custom domain and it worked, so I think that is indeed the case.

**Comment by brentvollebregt** [comment]
I've updated how this project is hosted - the frontend and API are now both at https://spotify-lyrics-viewer.nitratine.net/.

This seems to have resolved the iOS issue.

**Comment by botaoshen** [comment]
amazing work, can confirm it is working prefect.
